keps/sig-node/4205-psi-metric/README.md

```diff
@@ -0,0 +1,646 @@
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding support in kubelet to read Pressure Stall Information (PSI) metric pertaining to CPU, Memory and IO resources exposed from cAdvisor and runc. This will enable kubelet to report node conditions which will be utilized to prevent scheduling of pods on nodes experiencing significant resource constraints.
+
+## Motivation
+
+[PSI metric](https://www.kernel.org/doc/Documentation/accounting/psi.txt) provides a quantifiable way to see resource pressure increases as they develop, with a new pressure metric for three major resources (memory, CPU, IO). These pressure metrics are useful for detecting resource shortages and provide nodes the opportunity to respond intelligently - by updating the node condition.
+
+In short, PSI metric are like barometers that provide fair warning of impending resource shortages on the node, and enable nodes to take more proactive, granular and nuanced steps when major resources (memory, CPU, IO) start becoming scarce.
+
+### Goals
+
+This proposal aims to:
+1. Enable the kubelet to have the PSI metric of cgroupv2 exposed from cAdvisor and Runc.
+2. Enable the pod level PSI metric and expose it in the Summary API.
+3. Utilize the node level PSI metric to set node condition and node taints.
+
+It will have two phases:
+Phase 1: includes goal 1, 2 
+Phase 2: includes goal 3
+
+### Non-Goals
+
+* Invest in more opportunities to further use PSI metric for pod evictions,
+userspace OOM kills, and so on, for future KEPs.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+Today, to identity disruptions caused by resource crunches, Kubernetes users need to
```

Comment:
In might be worth to include that the node pressure conditions today are a bit limited [1]

The current node pressure conditions for memory pressure only verifies that a certain portion of memory is available, but not if there is memory pressure (and the thinking is that PSI is a more accurate representation of memory pressure). Also, there is no kubelet condition for cpu pressure or I/O pressure which we will gain with this KEP.

[1] https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-conditions